### PR TITLE
Command reports

### PR DESCRIPTION
**Pull request by Just-a-Unity-Dev**
## About the PR 
This PR adds Command Reports! A new toy for admins to play around with. It is simply an IC way of communicating with the Captain/Crew without using `announce`.

#### Mini Header: How to use
It's just like `announce` but better.

`sendcommandreport <classified (bool)> <message (string)>`

If `<classified>` is set to true, then it won't broadcast to the radio/entire station.

_Note that regardless of `classified`, it will be faxed._

**Screenshots**
image

image

**Changelog**

:cl: eclips_e
- add: Added Command Reports! Central Command can now fax messages privately to the bridge or broadcast it to the whole crew.

### CODE REVIEW

**Comment by Just-a-Unity-Dev** [comment]
ready for review

File: Content.Server/CommandReport/CommandReportSystem.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        /// <returns>True if at least one communication console got the report</returns>
+        public bool SendCommandReport(bool broadcastToRadio, string message)
+        {
+            // Such copypaste code... So sorry...
+            var wasSent = false;
+            var messageWrap = Loc.GetString("command-reports-sender-announcement-wrap-message");
+
+            SoundSystem.Play(AnnouncementSound, Filter.Broadcast(), AudioParams.Default.WithVolume(-2f));  // Play the "New Command Report Created" sound effect
+
+            _chatManager.ChatMessageToAll(ChatChannel.Radio, broadcastToRadio == true ? message : Loc.GetString("command-reports-confidential-announcement-message"), messageWrap, colorOverride: Color.Gold);
```
Comment (metalgearsloth) [inline Content.Server/CommandReport/CommandReportSystem.cs:36]: Why don't you just dispatchstationannouncement? You should probably also allow an optional station arg to be passed in.

File: Content.Server/CommandReport/CommandReportSystem.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                var paper = EntityManager.SpawnEntity("Paper", consolePos); // the paper that will be sent to the console
+
+                var papercomp = EntityManager.GetComponent<PaperComponent>(paper); // the PaperComponent of the paper
+                var papermeta = EntityManager.GetComponent<MetaDataComponent>(paper); // the metadata of the paper
+
+                papercomp.Content = $"Central Command Report\n{message}"; // header + contents
+                papermeta.EntityName = Loc.GetString("command-reports-paper-name"); // set the name of the paper to something special
+
+                wasSent = true; // fax has been sent at least once
+            }
```
Comment (metalgearsloth) [inline Content.Server/CommandReport/CommandReportSystem.cs:55]: wat.

1. Way too many comments.
2. Just add transformcomponent to the query
3. You should localise centcomm report.

File: Content.Server/CommandReport/Commands/SendCommandReport.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            if (args.Length < 2)
+            {
+                shell.WriteError("Not enough arguments! Need at least 2.");
+                return;
+            }
+
+            EntitySystem.Get<CommandReportSystem>().SendCommandReport(bool.Parse(args[0]), args[1]);
+
+            shell.WriteLine("Sent!");
+        }
```
Comment (metalgearsloth) [inline Content.Server/CommandReport/Commands/SendCommandReport.cs:36]: Localise. You should also specify that it's the contents that get reported or not to the radio.

File: Resources/Audio/Announcements/license.txt [hunk 4]
```diff
@@ -4,5 +4,6 @@ meteors.ogg taken from https://github.com/tgstation/tgstation/blob/95731342b9716
 outbreak7.ogg taken from /tg/station at commit https://github.com/tgstation/tgstation/commit/40d89d11ea4a5cb81d61dc1018b46f4e7d32c62a used under CC-BY-SA-3.0
 welcome.ogg taken from /tg/station at commit https://github.com/tgstation/tgstation/commit/40d89d11ea4a5cb81d61dc1018b46f4e7d32c62a used under CC-BY-SA-3.0
 announce.ogg taken from /tg/station https://github.com/tgstation/tgstation/commit/40d89d11ea4a5cb81d61dc1018b46f4e7d32c62a cut from the beginning of "shuttlerecalled.ogg" used under CC-BY-SA-3.0
+commandreport.ogg taken from at commit /tg/station/ https://github.com/tgstation/tgstation/commit/95731342b97167d7883ff091d389f79c36442ee6 used under CC-BY-SA-3.0
```
Comment (metalgearsloth) [inline Resources/Audio/Announcements/license.txt:7]: snake_case please.

File: Resources/Locale/en-US/commandreport/commandreports.ftl [hunk 5]
```diff
@@ -0,0 +1,5 @@
+﻿command-reports-sender-announcement-wrap-message = Central Command Update:
+                                                   {"{0}"}
```
Comment (metalgearsloth) [inline Resources/Locale/en-US/commandreport/commandreports.ftl:2]: Just newline it?

File: Resources/Locale/en-US/commandreport/commandreports.ftl [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+command-reports-confidential-announcement-message = A summary has been copied and printed at all communication consoles.
+command-reports-paper-name = paper - central command report
+command-reports-paper-description = If you see do eventually see a command report, it is either because you fucked up or the crew fucked up. Either way, CentComm is angry.
```
Comment (metalgearsloth) [inline Resources/Locale/en-US/commandreport/commandreports.ftl:5]: If you see do

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/CommandReport/CommandReportSystem.cs (hunk 1 above)
Comment (Just-a-Unity-Dev) [reply Content.Server/CommandReport/CommandReportSystem.cs:35]: > Why don't you just dispatchstationannouncement? You should probably also allow an optional station arg to be passed in.

DispatchStationAnnouncement has the `(Sender) Announcement`, but all I want is `(Sender)`

File: Content.Server/CommandReport/CommandReportSystem.cs (hunk 2 above)
Comment (Just-a-Unity-Dev) [reply Content.Server/CommandReport/CommandReportSystem.cs]: can you clarify on 2?

Comment (metalgearsloth) [reply Content.Server/CommandReport/CommandReportSystem.cs]: EntityQuery can take up to four generics.

File: Content.Server/CommandReport/CommandReportSystem.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    public class CommandReportSystem : EntitySystem
+    {
+        [Dependency] private readonly IChatManager _chatManager = default!;
+        [Dependency] private readonly IAdminLogManager _adminLogger = default!;
+        private const string AnnouncementSound = "/Audio/Announcements/command_report.ogg";
+        public override void Initialize()
+        {
+            base.Initialize();
+        }
+
```
Comment (mirrorcult) [inline Content.Server/CommandReport/CommandReportSystem.cs:23]: remove

File: Content.Server/CommandReport/CommandReportSystem.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        /// </summary>
+        /// <returns>True if at least one communication console got the report</returns>
+        public bool SendCommandReport(bool broadcastToRadio, string message)
+        {
+            var wasSent = false;
+            var messageWrap = Loc.GetString("command-reports-sender-announcement-wrap-message");
+
+            SoundSystem.Play(AnnouncementSound, Filter.Broadcast(), AudioParams.Default.WithVolume(-2f));  // Play the "New Command Report Created" sound effect
+
+            _chatManager.ChatMessageToAll(ChatChannel.Radio, broadcastToRadio == true ? message : Loc.GetString("command-reports-confidential-announcement-message"), messageWrap, colorOverride: Color.Gold);
```
Comment (mirrorcult) [inline Content.Server/CommandReport/CommandReportSystem.cs:35]: this doesn't work with multiple stations at all, it'll just send to one station. this needs to support taking in the station's `EntityUid` and use `DispatchStationAnnouncement` instead

File: Content.Server/CommandReport/Commands/SendCommandReport.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+using JetBrains.Annotations;
+using Robust.Shared.Audio;
+using Robust.Shared.Console;
+using Robust.Shared.Player;
+
+namespace Content.Server.CommandReport.Commands
+{
+    [UsedImplicitly]
+    [AdminCommand(AdminFlags.Fun)]
+    public sealed class SendNukeCodesCommand : IConsoleCommand
```
Comment (mirrorcult) [inline Content.Server/CommandReport/Commands/SendCommandReport.cs:19]: `SendNukeCodesCommand`?

File: Resources/Locale/en-US/commandreport/commandreports.ftl [hunk 10]
```diff
@@ -0,0 +1,9 @@
+﻿command-reports-sender-announcement-wrap-message = Central Command Update:
+{"{0}"}
+command-reports-confidential-announcement-message = A summary has been copied and printed at all communication consoles.
+command-reports-paper-name = paper - central command report
+command-reports-paper-content = Central Command Report
+{message}
+command-reports-paper-description = If you do eventually see a command report, it is either because you fucked up or the crew fucked up. Either way, CentComm is angry.
```
Comment (mirrorcult) [inline Resources/Locale/en-US/commandreport/commandreports.ftl:7]: this is never used?

File: Content.Server/CommandReport/CommandReportSystem.cs [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+            // todo: convert this to fax along with NukeCodeSystem.cs
```
Comment (mirrorcult) [inline Content.Server/CommandReport/CommandReportSystem.cs:37]: yeah i'm not sure about the duplication here. if this is added at all it needs to be deduplicated with the way `NukeCodeSystem` does it, not just added as a todo

**Review by mirrorcult — CHANGES_REQUESTED** [review]

**Comment by Just-a-Unity-Dev** [comment]
i'll work on this PR sooner or later since im busy this week

**Pull request closed without merging**